### PR TITLE
Respond to collisions

### LINKED ISSUE #1: Add simple collision response
The current code detects collisions (tinting the sprites red) but does not do anything about it.  In particular, the `calcCollisions` method has a `// ToDo` comment where the collision response should be.

This is where we need to call a new function that calculates and applies the collision forces on the bodies.

This is a well-known problem; searching for "2d physics collision response" should turn up some algorithms.  Books such as [_Physics for Game Developers_](https://www.amazon.com/Physics-Game-Developers-David-Bourg/dp/0596000065) are also a great reference.

For this ticket, it's fine if the collision code only works for the simple case of single-Bounds physics bodies.  We can generalize it to more complex shapes later.

We'll know the code is working when the test program (`test.ms`) shows the two small blocks coming to rest on the ground, one probably on top of the other (perhaps at a jaunty angle).

### PR DESCRIPTION
**Pull request by pipe01**
First time writing MiniScript, sorry for any mistakes!

In the `collisions.ms` file I've added a Separating Axis Theorem collision detection implementation, which allows you to find whether or not two objects are colliding as well as finding out how much they are overlapping by. To aid with this I've also implemented a simple Vector2 type to handle vector math more easily, it's contained in the `vector.ms` file. It's not super optimized yet, but I'm leaving that for later.

The collision response itself isn't done yet, what's there now is only a naive algorithm that always moves just one object. I would like to add some kind of force system, perhaps even with friction and bounciness.

I would also like to add an example for circles, which is one of the SAT algorithm's edge cases.

Fixes #1

### CODE REVIEW

**Comment by JoeStrout** [comment]
Hey that sounds like a great start!  The Rigidbody class already handles forces to some extent, though I don't think it's yet handling off-axis forces (imparting torque).  So that's something we will need.

I do worry a bit about the performance of using SAT on every pair of rigidbodies.  But I guess let's just get it working first, and we can optimize later.

A vec2 module is a good idea, however, instead of using maps, I would prefer to use a 2-element list — and we can define accessors to allow us to reference elements as .x and .y instead of [0] and [1], like this:
```
// accessors so you can use pos.x as a synonym for pos[0], etc.
list.x = function; return self[0]; end function
list.y = function; return self[1]; end function
```

But the nice thing about this is, you don't really need a separate factory function to make these; you can just return `[x+dx, y+dy]` or whatever, and this is a perfectly valid vector.  They also print more compactly and are probably more efficient than maps, too.

Welcome to MiniScript, and thanks for your efforts!  I look forward to working with you on this.

**Comment by pipe01** [comment]
> But the nice thing about this is, you don't really need a separate factory function to make these; you can just return [x+dx, y+dy] or whatever, and this is a perfectly valid vector. They also print more compactly and are probably more efficient than maps, too.

If I understand it correctly then that list wouldn't have any methods of the vector type right?. We could then instead make regular functions in the vector module, but that seems a bit more unwieldy. Although it quite handy to be able to create them with just a list, hmm..

**Comment by JoeStrout** [comment]
> If I understand it correctly then that list wouldn't have any methods of the vector type right?. We could then instead make regular functions in the vector module, but that seems a bit more unwieldy.

Not necessarily.  We can add more list extension methods, as needed.  For some examples (and to be aware of the list extensions the user will likely already have in effect), see: https://github.com/JoeStrout/minimicro-sysdisk/blob/master/sys/lib/listUtil.ms

One thing we _can't_ do in MiniScript is operator overloading.  But notice that the above module (which you can use in your own Mini Micro code by doing `import "listUtil"` at the top of your script) adds things like `list.add` and `list.plus`.  It even defines `list.dot` for doing a dot product.  So, much of what you need for a vector class is already there.

**Comment by pipe01** [comment]
That would work but then we would be polluting the list type with vector methods, for example the vector's `add` method would collide with list's `add` method.

**Comment by JoeStrout** [comment]
Also take a look at https://github.com/JoeStrout/minimicro-sysdisk/blob/master/sys/lib/matrixUtil.ms, which provides code for 2D matrices.  Not sure if we really need that for 2D physics, but it's available.

**Comment by JoeStrout** [comment]
> That would work but then we would be polluting the list type with vector methods, for example the vector's add method would collide with list's add method.

Sorry, I was unclear.  Let me try again:

_Use_ the standard listUtil method.  Just import it at the top of any physics script.  It's a standard system-supplied library, so always available.

Now you don't need to add a "vector `add`" method, because it's already there.  Same with multiply, dot, etc.  Everything you get for free in listUtil is already done.

We can add the `.x` and `.y` list extension methods, just to make code more readable (though it also makes it slower, so in tight loops we might want to just use `[0]` and `[1]` instead).  And if you need any _other_ vector-type list extension methods that aren't already in listUtil, add those in the new physics code.  I don't know off the top of my head what those might be, though; listUtil is pretty complete.

Feel free to join me on Discord if you want to chat more interactively!  Link's on the MiniScript home page.

**Comment by pipe01** [comment]
Ohh sorry, okay. I'm stupid and just saw the first couple of methods, I didn't see that there were more. Yeah I can totally use those and add more as I see fit.

**Comment by JoeStrout** [comment]
Thanks for your hard work!  It looks like all we need now is torques (so blocks will collide & rotate properly).
